posthog/api/event.py

```diff
@@ -267,122 +270,146 @@ def values(self, request: request.Request, **kwargs) -> response.Response:
-            values = []
-            for value in result.results:
-                if isinstance(value[0], float | int | bool | uuid.UUID):
+        order_by = []
+        if value:
+            order_by = [
+                ast.OrderExpr(
+                    expr=ast.Call(name="length", args=[ast.Call(name="toString", args=[ast.Field(chain=chain)])]),
+                    order="ASC",
+                )
+            ]
+        query = ast.SelectQuery(
+            select=[ast.Field(chain=chain)],
+            distinct=True,
+            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
+            where=ast.And(exprs=conditions),
+            order_by=order_by,
+            limit=ast.Constant(value=10),
+        )
+        result = execute_hogql_query(query, team=team)
+        values = []
+        for value in result.results:
+            if isinstance(value[0], float | int | bool | uuid.UUID):
+                values.append(value[0])
+            else:
+                try:
+                    values.append(json.loads(value[0]))
+                except json.JSONDecodeError:
                     values.append(value[0])
-                else:
-                    try:
-                        values.append(json.loads(value[0]))
-                    except json.JSONDecodeError:
-                        values.append(value[0])
-
         return response.Response([{"name": convert_property_value(value)} for value in flatten(values)])
 
+    @tracer.start_as_current_span("events_api_is_property_hidden")
+    def _is_property_hidden(self, key, team):
+        property_is_hidden = False
+        try:
+            from ee.models.property_definition import EnterprisePropertyDefinition
+
+            property_is_hidden = EnterprisePropertyDefinition.objects.filter(
+                team=team,
+                name=key,
+                type=PropertyDefinition.Type.EVENT.value,
+                hidden=True,
+            ).exists()
+        except ImportError:
+            # Enterprise features not available, continue normally
+            pass
+
+        return property_is_hidden
+
+    @tracer.start_as_current_span("events_api_custom_event_values")
+    def _custom_event_values(self, team: Team) -> response.Response:
+        system_events = [
+            event_name
+            for event_name in CORE_FILTER_DEFINITIONS_BY_GROUP["events"].keys()
+            if event_name != "All Events"  # Skip the wildcard
+        ]
+        query = ast.SelectQuery(
+            select=[ast.Field(chain=["event"])],
+            distinct=True,
+            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
+            where=ast.CompareOperation(
+                op=ast.CompareOperationOp.NotIn,
+                left=ast.Field(chain=["event"]),
+                right=ast.Constant(value=system_events),
+            ),
+            order_by=[ast.OrderExpr(expr=ast.Field(chain=["event"]), order="ASC")],
+        )
```

Comment:
there's no time limit on this query, no wonder it's slow for us
we should have a time limit here and let the user prompt for more in the UI
but baby steps